docs/src/main/paradox/server-side/websocket-support.md

```diff
@@ -46,28 +46,55 @@ For sending data, you can use @scala[`TextMessage.apply(text: String)`]@java[`Te
 to create a streaming message from an Akka Stream source.
 
-## Server API
+## Routing support
 
-The entrypoint for the WebSocket API is the synthetic @apidoc[UpgradeToWebSocket] header which is added to a request
-if Akka HTTP encounters a WebSocket upgrade request.
+To handle websocket requests, you can either use the directive described in this section, or
+use the more low-level @ref[WebSocketUpgrade](#websocketupgrade) attribute described
+in the next section.
 
-The WebSocket specification mandates that details of the WebSocket connection are negotiated by placing special-purpose
-HTTP-headers into request and response of the HTTP upgrade. In Akka HTTP these HTTP-level details of the WebSocket
-handshake are hidden from the application and don't need to be managed manually.
+The routing DSL provides the @ref[handleWebSocketMessages](../routing-dsl/directives/websocket-directives/handleWebSocketMessages.md) directive to install a WebSocket handler if a request
+is a WebSocket request. Otherwise, the directive rejects the request.
 
-@@@ note
+Let's look at how the above example can be rewritten using the high-level routing DSL.
+
+Instead of writing the request handler manually, the routing behavior of the app is defined by a route that
+uses the @ref[handleWebSocketMessages](../routing-dsl/directives/websocket-directives/handleWebSocketMessages.md) directive:
 
-Prefer using the @ref[Routing Support](#routing-support) as this implementation with a synthetic headers will be replaced with using request attributes (see [issue #3278](https://github.com/akka/akka-http/issues/3278)). 
+Scala
+:  @@snip [WebSocketDirectivesExamplesSpec.scala](/docs/src/test/scala/docs/http/scaladsl/server/directives/WebSocketDirectivesExamplesSpec.scala) { #greeter-service }
 
+Java
+:  @@snip [WebSocketRoutingExample.java](/docs/src/test/java/docs/http/javadsl/server/WebSocketRoutingExample.java) { #websocket-route }
+
+The handling code itself will be the same as with using the low-level API.
+
+@@@ div { .group-scala }
+The example also includes code demonstrating the testkit support for WebSocket services. It allows to create WebSocket
+requests to run against a route using *WS* which can be used to provide a mock WebSocket probe that allows manual
+testing of the WebSocket handler's behavior if the request was accepted.
 @@@
 
-Instead, the synthetic @apidoc[UpgradeToWebSocket] represents a valid WebSocket upgrade request. An application can detect
-a WebSocket upgrade request by looking for the @apidoc[UpgradeToWebSocket] header. It can choose to accept the upgrade and
+See the @github[full routing example](/docs/src/test/java/docs/http/javadsl/server/WebSocketCoreExample.java).
+
+## WebSocketUpgrade
+
+To handle websocket requests, you can either use the @apidoc[WebSocketUpgrade] attribute directly, or
+use the more high-level @ref[Routing Support](#routing-support) described below.
```

Comment:
```suggestion
use the more high-level @ref[Routing Support](#routing-support) described above.
```

Reply:
Heh. I actually planned to remove this paragraph (and forgot), but keeping it like this is even better :+1: